sdk/plugin/plugin_v5.go

```diff
@@ -0,0 +1,165 @@
+package plugin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	plugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/hashicorp/vault/sdk/helper/pluginutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/sdk/plugin/pb"
+)
+
+// BackendPluginClientV5 is a wrapper around backendPluginClient
+// that also contains its plugin.Client instance. It's primarily
+// used to cleanly kill the client on Cleanup()
+type BackendPluginClientV5 struct {
+	client pluginutil.PluginClient
+
+	logical.Backend
+}
+
+type ContextKey string
+
+func (c ContextKey) String() string {
+	return "plugin" + string(c)
+}
+
+const ContextKeyPluginReload = ContextKey("plugin-reload")
+
+// Cleanup cleans up the go-plugin client and the plugin catalog
+func (b *BackendPluginClientV5) Cleanup(ctx context.Context) {
+	_, ok := ctx.Value(ContextKeyPluginReload).(string)
+	if !ok {
+		b.Backend.Cleanup(ctx)
+		b.client.Close()
+		return
+	}
+	b.Backend.Cleanup(ctx)
+	b.client.Reload()
+}
+
+// NewBackendV5 will return an instance of an RPC-based client implementation of
+// the backend for external plugins, or a concrete implementation of the
+// backend if it is a builtin backend. The backend is returned as a
+// logical.Backend interface.
+func NewBackendV5(ctx context.Context, pluginName string, pluginType consts.PluginType, sys pluginutil.LookRunnerUtil, conf *logical.BackendConfig) (logical.Backend, error) {
+	// Look for plugin in the plugin catalog
+	pluginRunner, err := sys.LookupPlugin(ctx, pluginName, pluginType)
+	if err != nil {
+		return nil, err
+	}
+
+	var backend logical.Backend
+	if pluginRunner.Builtin {
+		// Plugin is builtin so we can retrieve an instance of the interface
+		// from the pluginRunner. Then cast it to logical.Factory.
+		rawFactory, err := pluginRunner.BuiltinFactory()
+		if err != nil {
+			return nil, fmt.Errorf("error getting plugin type: %q", err)
+		}
+
+		if factory, ok := rawFactory.(logical.Factory); !ok {
+			return nil, fmt.Errorf("unsupported backend type: %q", pluginName)
+		} else {
```

Comment:
nit: else not needed here since if block above returns

Reply:
I am going to leave this as is since `factory` is just scoped to this if/else clause and removing the else would require more refactoring.